### PR TITLE
Filter templates & snapshots by store id

### PR DESCRIPTION
**Pull request by vishesh92**
### Description

This PR adds store's details in template & snapshot response and allows filtering on their IDs as well.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: api/src/main/java/org/apache/cloudstack/api/command/admin/snapshot/CreateSnapshotFromVMSnapshotCmdByAdmin.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import com.cloud.storage.Snapshot;
+import org.apache.cloudstack.api.APICommand;
+import org.apache.cloudstack.api.ResponseObject;
+import org.apache.cloudstack.api.command.admin.AdminCmd;
+import org.apache.cloudstack.api.command.user.snapshot.CreateSnapshotFromVMSnapshotCmd;
+import org.apache.cloudstack.api.response.SnapshotResponse;
+
+@APICommand(name = "createSnapshotFromVMSnapshot",
+            description = "Creates an instant snapshot of a volume from existing vm snapshot.",
+            responseObject = SnapshotResponse.class, entityType = {Snapshot.class}, since = "4.10.0",
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/snapshot/CreateSnapshotFromVMSnapshotCmdByAdmin.java:28]:
```suggestion
            responseObject = SnapshotResponse.class, entityType = {Snapshot.class},
```

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java [hunk 2]
```diff
@@ -672,17 +672,25 @@ public SnapshotResponse createSnapshotResponse(Snapshot snapshot) {
             snapshotInfo = snapshotfactory.getSnapshot(snapshot.getId(), dataStoreRole);
         }
 
+        if (view.equals(ResponseView.Full)) {
+            SnapshotDataStoreVO snapshotStore = _snapshotStoreDao.findBySnapshot(snapshot.getId(), DataStoreRole.Image);
+            long storagePoolId = snapshotStore.getDataStoreId();
+            DataStore dataStore = _dataStoreMgr.getDataStore(storagePoolId, DataStoreRole.Image);
+            snapshotResponse.setStore(dataStore.getName());
+            snapshotResponse.setStoreId(dataStore.getUuid());
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/api/ApiResponseHelper.java:681]: can be a separate method

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 3]
```diff
@@ -3639,8 +3646,25 @@ private Pair<List<TemplateJoinVO>, Integer> searchForTemplatesInternal(ListTempl
...
+
+            if (templatesInStore == null || templatesInStore.isEmpty()) {
+                return new Pair<>(new ArrayList<>(), 0);
+            }
+            if (ids.isEmpty()) {
+                ids = templatesInStore.stream().map(TemplateDataStoreVO::getTemplateId).collect(Collectors.toList());
+            } else {
+                ids.retainAll(templatesInStore.stream().map(TemplateDataStoreVO::getTemplateId).collect(Collectors.toList()));
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3662]: can this be a separate method as well?

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java [hunk 4]
```diff
@@ -685,6 +686,21 @@ public Pair<List<? extends Snapshot>, Integer> listSnapshots(ListSnapshotsCmd cm
...
+            if (snapshotsInStore == null || snapshotsInStore.isEmpty()) {
+                return new Pair<>(new ArrayList<>(), 0);
+            }
+            List<Long> snapshotIdsInStore = snapshotsInStore.stream().map(SnapshotDataStoreVO::getSnapshotId).collect(Collectors.toList());
+            if (ids.isEmpty()) {
+                ids = snapshotIdsInStore;
+            } else {
+                ids.retainAll(snapshotIdsInStore);
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:702]: safe the poorly placed return, this can be separate method

**Comment by vishesh92** [comment]
@blueorangutan package

**Review by kiranchavala — COMMENTED** [review]
@vishesh92 

In the UI under storage>snapshots ,

I see the label mentioned, could you please make it Imagestore

listsnapshots

Similarly for Images> templates > can you add a column for image store 

list templates

**Comment by kiranchavala** [comment]
@vishesh92 can this PR  be extended out for " list iso" also as there are present in the secondary storage

https://cloudstack.apache.org/api/apidocs-4.18/apis/listIsos.html

**Comment by vishesh92** [comment]
> @vishesh92
> 
> In the UI under storage>snapshots ,
> 
> I see the label mentioned, could you please make it Imagestore
> 
> listsnapshots
> 
> Similarly for Images> templates > can you add a column for image store
> 
> list templates

I am getting this. I think you checked an older commit.
image

**Comment by vishesh92** [comment]
> @vishesh92 can this PR be extended out for " list iso" also as there are present in the secondary storage
> 
> https://cloudstack.apache.org/api/apidocs-4.18/apis/listIsos.html

@kiranchavala Let me include ISOs as well.

**Comment by vishesh92** [comment]
Continuing this work in https://github.com/apache/cloudstack/pull/7949

**Pull request closed without merging**